### PR TITLE
Implements new indicator ForceIndex

### LINKED ISSUE #8123: Implement Force Index
#### Expected Behavior
Supports Force Index (see [talipp implementation](https://github.com/nardew/talipp/blob/main/talipp/indicators/ForceIndex.py)).
It's a moving average of the return (1-period price difference) multiplied by the volume.

Refer to [LEAN Engine > Contributions > Indicators](https://www.quantconnect.com/docs/v2/lean-engine/contributions/indicators).

#### Checklist
- [x] I have completely filled out this template
- [x] I have confirmed that this issue exists on the current `master` branch
- [x] I have confirmed that this is not a duplicate issue by searching [issues](https://github.com/QuantConnect/Lean/issues)
- [x] I have provided detailed steps to reproduce the issue

### PR DESCRIPTION
**Pull request by femtotrader**
#### Description

Implement new indicator - ForceIndex

#### Related Issue

Closes #8123

#### Motivation and Context
The ForceIndex indicator wasn't implemented

#### Requires Documentation Change

Doc for this indicator should be created but some docstring are written

#### How Has This Been Tested?

comparison with talipp computation of ForceIndex(20) of SPY candle data using [generate_reference_data_from_talipp.py](https://github.com/QuantConnect/Lean/blob/master/Tests/TestData/generate_reference_data_from_talipp.py)

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

**Comment by femtotrader** [comment]
I have no idea why AcceptsRenkoBarsAsInput test is failing.
All others tests are passing

**Comment by jaredbroad** [comment]
Thanks, @femtotrader! For all the new PRs, can you please attach the source code of the comparison with Tulip, which generates the test data? We want to make sure all indicators have a third-party benchmark.

**Comment by femtotrader** [comment]
Indicators are tested against talipp (Python) not tulipindicators (which is a great technical indicator lib also... but in C)

Source code of the comparison with talipp have ever been merged in this PR
https://github.com/QuantConnect/Lean/pull/8148/
see generate_reference_data_from_talipp.py

I don't have an other (second) comparison script for tulipindicators .

**Comment by femtotrader** [comment]
Tulip indicators have a Python binding https://github.com/TulipCharts/tulipy which is marked as [NOT ACTIVELY MAINTAINED]  ... so I don't know what to think about it.

This fork https://github.com/jesse-ai/tulipy (which have been used by Jesse bot) doesn't seems to be in a better state.

If comparison should be done against tulip indicators also we probably should rely on a C/C++ program for that purpose (which is an other story)

File: Indicators/ForceIndex.cs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        protected override decimal ComputeNextValue(TradeBar input)
+        {
+            if (Samples < 2)
+            {
+                _previousInput = input;
+                return 0;
+            }
+            // compute the instantaneous force index and then send it to our smoother
+            
+            _smoother.Update(input.Time, (input.Close - _previousInput.Close) * input.Volume);
```
Comment (Martin-Molinero) [inline Indicators/ForceIndex.cs:83]: Minor comment, we should use `input.EndTime` here, which is the current time, bars go from Time to EndTime

File: Indicators/ForceIndex.cs [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+ * limitations under the License.
+*/
+
+using QuantConnect.Data.Market;
+
+namespace QuantConnect.Indicators
+{
+    /// <summary>
+    ///The Force Index is calculated by comparing the current market price with the previous market price 
+    ///and multiplying its difference with the traded volume during a specific time period.
```
Comment (Martin-Molinero) [inline Indicators/ForceIndex.cs:22]: nit - missing a space for xml docs

File: Indicators/ForceIndex.cs [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    ///
+    /// </summary>
+    public class ForceIndex : TradeBarIndicator, IIndicatorWarmUpPeriodProvider
+    {
+        private TradeBar _previousInput;
+
+        /// <summary>This indicator is used to smooth the ForceIndex computation</summary>
```
Comment (Martin-Molinero) [inline Indicators/ForceIndex.cs:29]: Summary tags are usually in their own lines, see other xml docs

**Review by Martin-Molinero — CHANGES_REQUESTED** [review]
Nice! Looks good, sharing minor requests 👍

File: Tests/Indicators/ForceIndexTests.cs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+using System;
+using NUnit.Framework;
+using QuantConnect.Data.Market;
+using QuantConnect.Indicators;
+
+namespace QuantConnect.Tests.Indicators
+{
+    [TestFixture]
+    public class ForceIndexTests : CommonIndicatorTests<TradeBar>
```
Comment (Martin-Molinero) [inline Tests/Indicators/ForceIndexTests.cs:24]: Some tests seems to be failing
image

File: Indicators/ForceIndex.cs (hunk 3 above)
Comment (femtotrader) [reply Indicators/ForceIndex.cs]: Fixed by https://github.com/QuantConnect/Lean/pull/8155/commits/1ecec6636fb83897701ec0c55a938f6591f077a9

File: Indicators/ForceIndex.cs (hunk 2 above)
Comment (femtotrader) [reply Indicators/ForceIndex.cs]: Fixed by https://github.com/QuantConnect/Lean/pull/8155/commits/1ecec6636fb83897701ec0c55a938f6591f077a9

File: Indicators/ForceIndex.cs (hunk 1 above)
Comment (femtotrader) [reply Indicators/ForceIndex.cs]: Fixed by https://github.com/QuantConnect/Lean/pull/8155/commits/1ecec6636fb83897701ec0c55a938f6591f077a9

File: Tests/Indicators/ForceIndexTests.cs (hunk 4 above)
Comment (femtotrader) [reply Tests/Indicators/ForceIndexTests.cs:24]: Well this one is more complicated and will probably need some investigation from your side.

I encountered similar problem with some others PR.

That's why you you see https://github.com/QuantConnect/Lean/pulls/femtotrader Tasks 5 of 11 and not 6 of 11

My guess is that either I didn't understand what happen when indicator have Renko Bars as input (if this kind of indicator can really take them) or if I should short-circuit test or if I need to see if the problem is on AcceptsRenkoBarsAsInput code?

Anyway some help / guidance will help as it's locking

Comment (Martin-Molinero) [reply Tests/Indicators/ForceIndexTests.cs:24]: Suggest debug the failing test and underlying why it's failing, some of these method in the base indicator test suite are virtual so they can be overwritten if required, u can see there are other couple of implementations of AcceptsRenkoBarsAsInput. But first should validate/understand what's going on, to know what's the solution 👍

Comment (femtotrader) [reply Tests/Indicators/ForceIndexTests.cs:24]: I won't be available more today neither tomorrow. Feel free to take the lead on this one (it should help others waiting PR)

**Review by Martin-Molinero — CHANGES_REQUESTED** [review]
Thank you @femtotrader! Almost there 👍 minor comment

**Review by Martin-Molinero — APPROVED** [review]

**Pull request merged**